state/cloudcredentials.go

```diff
@@ -17,6 +17,12 @@ import (
 	"github.com/juju/juju/cloud"
 )
 
+// Credential contains information about the credential as stored on
+// the controller.
+type Credential struct {
```

Comment:
I'm confused as to why we need a new type for Credential.
Won't cloud.Credential do the job?
Introducing a new type here means that we now have a mix of state.Credential and cloud.Credential.
Why would AllCloudCredentials() return state.Credential whereas CloudCredential() returns cloud.Credential?

Reply:
cloud.Credential does not work for several reasons.
(1) it does not differentiate between different types of attributes;
(2) it only has accessors and does not have a constructor that will take several attributes we are interested in;
(3) It has parse auth type from string into a typed property;
(4) It does not have other properties that we need like cloud, owner, etc.

I have pondered and may be convinced to have another type in juju/juju/cloud package that is more comprehensive, say cloud.FullCredential... It feels like mostly copying things from one place to another though and (2) will still apply (I personally dislike constructors that take millions things). 

What are your concrete objections to state.Credential as proposed?

Reply:
As discussed off-line, I have changed the original, existing state calls to return state.Credential instead of cloud.Credential.